tests/standalone_activity_test.go

```diff
@@ -105,3 +117,377 @@ func createDefaultInput() *commonpb.Payloads {
+	require.NotNil(t, firstPollResp.StateChangeLongPollToken)
+	require.NotNil(t, firstPollResp.Info)
+	s.assertActivityExecutionInfo(
+		t,
+		firstPollResp.Info,
+		activityID,
+		startResp.RunId,
+		enumspb.PENDING_ACTIVITY_STATE_SCHEDULED,
+	)
+
+	taskQueuePollErr := make(chan error, 1)
+	activityPollDone := make(chan struct{})
+	var activityPollResp *workflowservice.PollActivityExecutionResponse
+	var activityPollErr error
+
+	go func() {
+		defer close(activityPollDone)
+		// Second poll uses token and therefore waits for a state transition
+		activityPollResp, activityPollErr = s.FrontendClient().PollActivityExecution(ctx, &workflowservice.PollActivityExecutionRequest{
+			Namespace:    s.Namespace().String(),
+			ActivityId:   activityID,
+			RunId:        startResp.RunId,
+			IncludeInfo:  true,
+			IncludeInput: true,
+			WaitPolicy: &workflowservice.PollActivityExecutionRequest_WaitAnyStateChange{
+				WaitAnyStateChange: &workflowservice.PollActivityExecutionRequest_StateChangeWaitOptions{
+					LongPollToken: firstPollResp.StateChangeLongPollToken,
+				},
+			},
+		})
+	}()
+
+	// TODO(dan): race here: subscription might not be established yet
+
+	// Worker picks up activity task, triggering transition (via RecordActivityTaskStarted)
+	go func() {
+		_, err := s.pollActivityTaskQueue(ctx, taskQueue)
+		taskQueuePollErr <- err
+	}()
+
+	select {
+	case <-activityPollDone:
+		require.NoError(t, activityPollErr)
+		require.NotNil(t, activityPollResp)
+		require.NotNil(t, activityPollResp.Info)
+		s.assertActivityExecutionInfo(
+			t,
+			activityPollResp.Info,
+			activityID,
+			startResp.RunId,
+			enumspb.PENDING_ACTIVITY_STATE_STARTED,
+		)
+		require.NotNil(t, activityPollResp.Input)
+		require.Equal(t, "test-activity-input", string(activityPollResp.Input.Payloads[0].Data))
+
+	case <-ctx.Done():
+		t.Fatal("PollActivityExecution timed out")
+	}
+
+	err = <-taskQueuePollErr
+	require.NoError(t, err)
+
+	// Manipulate the token to verify token staleness checks (simulate ErrStaleReference). To do so
+	// we make use of the internal implementation detail that the bytes are a serialized ref.
+	token := firstPollResp.StateChangeLongPollToken
+	var pRef persistencespb.ChasmComponentRef
+	err = pRef.Unmarshal(token)
+	require.NoError(t, err)
+	if pRef.EntityVersionedTransition != nil {
+		pRef.EntityVersionedTransition.NamespaceFailoverVersion += 1
+	}
+	token, err = pRef.Marshal()
+	require.NoError(t, err)
+
+	_, err = s.FrontendClient().PollActivityExecution(ctx, &workflowservice.PollActivityExecutionRequest{
+		Namespace:  s.Namespace().String(),
+		ActivityId: activityID,
+		RunId:      startResp.RunId,
+		WaitPolicy: &workflowservice.PollActivityExecutionRequest_WaitAnyStateChange{
+			WaitAnyStateChange: &workflowservice.PollActivityExecutionRequest_StateChangeWaitOptions{
+				LongPollToken: token,
+			},
+		},
+	})
+	require.Error(t, err)
+	require.ErrorContains(t, err, "cached mutable state could potentially be stale")
```

Comment:
This should not be a user facing error. I would be scared as a user to see this come back.

Reply:
You should return some form of retryable error, probably one of our errors that map to an unavailable grpc status code. This can happen when a namespace fails over and replication hasn't caught up yet. The user's request should be allowed to retry as long as they are willing to.

Reply:
Just note that this counts as an SLA violation in Temporal cloud but I think that's expected.